policy/scan/pdque/queue.go

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package pdque
+
+import (
+	"bytes"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/spf13/afero"
+)
+
+const (
+	jobFileExt     = ".job"
+	tempFileExt    = ".tmp"
+	tempFilePrefix = "."
+)
+
+// ErrQueueClosed is the error returned when a queue is closed.
+var ErrQueueClosed = errors.New("queue is closed")
+
+var ErrQueueFull = errors.New("queue is full")
+
+// ErrUnableToDecode is returned when an object cannot be decoded.
+type ErrUnableToDecode struct {
+	Path string
+	Err  error
+}
+
+func (e ErrUnableToDecode) Error() string {
+	return fmt.Sprintf("object in file %s cannot be decoded: %s", e.Path, e.Err)
+}
+
+type Queue struct {
+	Name    string
+	path    string
+	mu      sync.Mutex
+	closed  bool
+	cond    *sync.Cond
+	builder func() interface{}
+	maxSize int
+}
+
+func New(name string, path string, maxSize int, builder func() interface{}) (*Queue, error) {
+	overlyPermissive, err := isOverlyPermissive(path)
+	if err != nil {
+		return nil, err
+	}
+	if overlyPermissive {
+		return nil, errors.New("path is overly permissive, make sure it is not writable to others or the group: " + path)
+	}
+
+	que := &Queue{
+		Name:    name,
+		path:    path,
+		builder: builder,
+		maxSize: maxSize,
+	}
+	que.cond = sync.NewCond(&que.mu)
+
+	return que, nil
+}
+
+func NewOrOpen(name string, path string, maxSize int, builder func() interface{}) (*Queue, error) {
```

Comment:
semantic thingy: The difference between `New` and `NewOrOpen` when the folder does not exist is that `NewOrOpen` will create the folder, while `New` will not. This feels unintuitive, since in the case where the folder doesn't exist both should behave similarly. 

As a possible suggestion if that makes sense, either just rename it to `CreateOrOpen` (making it different from `New` semantically) or implement the folder-check+create in `New` too.

Reply:
Oh, yeah `New` should of course create the directory, not just `NewOrOpen`